aten/src/ATen/cuda/CUDABlas.h

```diff
@@ -0,0 +1,251 @@
+  available in at::cuda::blas namespace.
+ */
+
+#include <ATen/cuda/CUDAContext.h>
+
+// In CUDA 8.0, definition of data types for sgemmex changed
+#if CUDA_VERSION < 8000
+#define CUDA_R_16F CUBLAS_DATA_HALF
+#endif
+
+#define TORCH_CUDABLAS_CHECK(EXPR)        \
+  do {                                    \
+    cublasStatus_t __err = EXPR;          \
+    if (__err != CUBLAS_STATUS_SUCCESS) { \
+      AT_ERROR(                           \
+          "CUDA error: ",                 \
+          _cublasGetErrorEnum(__err),     \
+          " when calling `" #EXPR "`");   \
+    }                                     \
+  } while (0)
+
+namespace {
+static const char* _cublasGetErrorEnum(cublasStatus_t error) {
+  if (error == CUBLAS_STATUS_SUCCESS) {
+    return "CUBLAS_STATUS_SUCCESS";
+  }
+  if (error == CUBLAS_STATUS_NOT_INITIALIZED) {
+    return "CUBLAS_STATUS_NOT_INITIALIZED";
+  }
+  if (error == CUBLAS_STATUS_ALLOC_FAILED) {
+    return "CUBLAS_STATUS_ALLOC_FAILED";
+  }
+  if (error == CUBLAS_STATUS_INVALID_VALUE) {
+    return "CUBLAS_STATUS_INVALID_VALUE";
+  }
+  if (error == CUBLAS_STATUS_ARCH_MISMATCH) {
+    return "CUBLAS_STATUS_ARCH_MISMATCH";
+  }
+  if (error == CUBLAS_STATUS_MAPPING_ERROR) {
+    return "CUBLAS_STATUS_MAPPING_ERROR";
+  }
+  if (error == CUBLAS_STATUS_EXECUTION_FAILED) {
+    return "CUBLAS_STATUS_EXECUTION_FAILED";
+  }
+  if (error == CUBLAS_STATUS_INTERNAL_ERROR) {
+    return "CUBLAS_STATUS_INTERNAL_ERROR";
+  }
+#if CUDA_VERSION >= 6000
+  if (error == CUBLAS_STATUS_NOT_SUPPORTED) {
+    return "CUBLAS_STATUS_NOT_SUPPORTED";
+  }
+#endif
+#if CUDA_VERSION >= 6050
+  if (error == CUBLAS_STATUS_LICENSE_ERROR) {
+    return "CUBLAS_STATUS_LICENSE_ERROR";
+  }
+#endif
+  return "<unknown>";
+}
+
+static cublasOperation_t _cublasOpFromChar(char op) {
+  switch (op) {
+    case 'n':
+      return CUBLAS_OP_N;
+    case 't':
+      return CUBLAS_OP_T;
+    case 'c':
+      return CUBLAS_OP_C;
+  }
+  AT_ERROR(
+      "_cublasOpFromChar input should be 't', 'n' or 'c' but got `", op, "`");
+}
+} // anonymous namespace
+
+namespace at {
+namespace cuda {
+namespace blas {
+
+#define GEMM_ARGTYPES(Dtype)                                               \
+  cudaStream_t stream, char transa, char transb, int64_t m, int64_t n,     \
+      int64_t k, Dtype alpha, const Dtype *a, int64_t lda, const Dtype *b, \
+      int64_t ldb, Dtype beta, Dtype *c, int64_t ldc
+
+template <typename Dtype>
+void gemm(GEMM_ARGTYPES(Dtype)) {
+  AT_ERROR("at::cuda::gemm: not implemented for ", typeid(Dtype).name());
+}
+
+template <>
+void gemm<double>(GEMM_ARGTYPES(double)) {
+  cublasHandle_t handle = at::cuda::getCurrentCUDABlasHandle();
+  cublasOperation_t opa = _cublasOpFromChar(transa);
+  cublasOperation_t opb = _cublasOpFromChar(transb);
+  TORCH_CUDABLAS_CHECK(cublasSetStream(handle, stream));
+  TORCH_CUDABLAS_CHECK(cublasDgemm(
+      handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc));
```

Comment:
It seems like we lost 32-bit checking for sizes here?

(Also, this code too would be pretty nice to deduplicate with the `THCBlas.cu` it was ported from, so we don't have to fix bugs in two locations.)

Reply:
Similar admonitions apply below.

Reply:
The 32-bit checking is now implemented. Issue #21566 is created to complete the CUDA Blas porting.